Index: git/unoconv
===================================================================
--- git.orig/unoconv	2021-09-28 17:06:37.559001448 +0200
+++ git/unoconv	2021-09-28 17:09:07.378878149 +0200
@@ -844,6 +844,7 @@
         global ooproc, product, office
         unocontext = None
 
+        info(3, "Resolving with 'uno:%s'" % op.connection)
         try:
             unocontext = resolver.resolve("uno:%s" % op.connection)
         except NoConnectException as e:
@@ -864,10 +865,10 @@
                 if op.userProfile:
                     args.append("-env:UserInstallation=file://" + realpath(op.userProfile))
                 info(2, '%s listener arguments are %s.' % (product.ooName, args))
-                ooproc = subprocess.Popen(args, env=os.environ)
+                ooproc = subprocess.Popen(args, env=os.environ, stdout=subprocess.PIPE)
                 info(2, '%s listener successfully started. (pid=%s)' % (product.ooName, ooproc.pid))
 
-                ### Try connection to it for op.timeout seconds (flakky OpenOffice)
+                ### Try connection to it for op.timeout seconds (flaky OpenOffice)
                 timeout = 0
                 while timeout <= op.timeout:
                     ### Is it already/still running ?
@@ -879,6 +880,20 @@
 
                     elif retcode != None:
                         info(3, "Process %s (pid=%s) exited with %s." % (office.binary, ooproc.pid, retcode))
+
+                        while True:
+                            output = ooproc.stdout.readline()
+                            if output:
+                                print ("Out: ",output.strip())
+                            else:
+                                print ("No output")
+                                break
+
+                        if retcode == 0:
+                            # Retry if the started listener exited too quickly, see https://github.com/unoconv/unoconv/issues/470
+                            print("Retrying as ooprocess stopped immediately")
+                            return self.connect(resolver)
+
                         break
 
                     try:
@@ -1287,7 +1302,7 @@
         if convertor.desktop.getCurrentFrame():
             info(2, 'Trying to stop %s GUI listener.' % product.ooName)
             try:
-                if product.ooName != "LibreOffice" or product.ooSetupVersion <= 3.3:
+                if product.ooName != "LibreOffice" or LooseVersion(product.ooSetupVersion) <= LooseVersion('3.3'):
                     subprocess.Popen([office.binary, "-headless", "-invisible", "-nocrashreport", "-nodefault", "-nofirststartwizard", "-nologo", "-norestore", "-unaccept=%s" % op.connection], env=os.environ)
                 else:
                     subprocess.Popen([office.binary, "--headless", "--invisible", "--nocrashreport", "--nodefault", "--nofirststartwizard", "--nologo", "--norestore", "--unaccept=%s" % op.connection], env=os.environ)
